codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecoder.java

```diff
@@ -73,76 +72,73 @@ protected void decode(ChannelHandlerContext ctx, HttpObject msg, List<Object> ou
+            } else {
+                contentEncoding = IDENTITY;
+            }
+            decoder = newContentDecoder(contentEncoding);
 
-                if (c instanceof LastHttpContent) {
-                    decodeStarted = false;
+            if (decoder == null) {
+                if (message instanceof HttpContent) {
+                    ((HttpContent) message).retain();
                 }
                 out.add(message);
-                out.add(c.retain());
                 return;
             }
 
-            if (decoder != null) {
-                decodeContent(c, out);
+            // Remove content-length header:
+            // the correct value can be set only after all chunks are processed/decoded.
+            // If buffering is not an issue, add HttpObjectAggregator down the chain, it will set the header.
+            // Otherwise, rely on LastHttpContent message.
+            headers.remove(HttpHeaderNames.CONTENT_LENGTH);
+
+            // set new content encoding,
+            CharSequence targetContentEncoding = getTargetContentEncoding(contentEncoding);
+            if (HttpHeaderValues.IDENTITY.equals(targetContentEncoding)) {
+                // Do NOT set the 'Content-Encoding' header if the target encoding is 'identity'
+                // as per: http://tools.ietf.org/html/rfc2616#section-14.11
+                headers.remove(HttpHeaderNames.CONTENT_ENCODING);
             } else {
-                if (c instanceof LastHttpContent) {
-                    decodeStarted = false;
+                headers.set(HttpHeaderNames.CONTENT_ENCODING, targetContentEncoding);
+            }
+
+            if (message instanceof HttpContent) {
+                // If message is a full request or response object (headers + data), don't copy data part into out.
+                // Output headers only; data part will be decoded below.
+                // Note: "copy" object must not be an instance of LastHttpContent class,
+                // as this would (erroneously) indicate the end of the HttpMessage to other handlers.
+                HttpMessage copy;
+                if (message instanceof HttpRequest) {
+                    HttpRequest r = (HttpRequest) message; // HttpRequest or FullHttpRequest
+                    copy = new DefaultHttpRequest(r.protocolVersion(), r.method(), r.uri());
+                } else if (message instanceof HttpResponse) {
+                    HttpResponse r = (HttpResponse) message; // HttpResponse or FullHttpResponse
+                    copy = new DefaultHttpResponse(r.protocolVersion(), r.status());
+                } else {
+                    throw new CodecException("Object of class " + message.getClass().getName() +
+                                             " is not a HttpRequest or HttpResponse");
                 }
+                copy.headers().set(message.headers());
+                copy.setDecoderResult(message.decoderResult());
```

Comment:
Scott, this is indeed a special copy with semantic "keep headers but drop content and drop flag of last message, if any", but I don't think a new method is justified.
First, HttpObject is an interface and adding a method would break all implementing classes, both netty's code (that's easy to fix) and 3rd-party ones (and they are out of control). Had HttpObject been an (abstract) class, then adding a method could be safe, but public interfaces, once published, are better stay immutable.
Second, this special copy is needed only once, here, and for a special case only - when HttpObjectAggregator precedes HttpContentDecoder in chain.
In total, I think it doesn't worth it.